src/sentry/rules/processing/delayed_processing.py

```diff
@@ -226,5 +274,30 @@ def apply_delayed(project_id: int) -> DefaultDict[Rule, set[int]] | None:
         rules_to_fire = get_rules_to_fire(
             condition_group_results, rule_to_slow_conditions, rules_to_groups
         )
-        return rules_to_fire
-    return None
+    # Step 7: Fire the rule's actions
+    now = timezone.now()
+    for rule, group_ids in rules_to_fire.items():
+        frequency = rule.data.get("frequency") or Rule.DEFAULT_FREQUENCY
+        freq_offset = now - timedelta(minutes=frequency)
+        group_to_groupevent = get_group_to_groupevent(rulegroup_to_events, project.id, group_ids)
+        for group, groupevent in group_to_groupevent.items():
+            rule_statuses = bulk_get_rule_status(alert_rules, group, project)
+            status = rule_statuses[rule.id]
+            if status.last_active and status.last_active > freq_offset:
+                return
+
+            updated = (
+                GroupRuleStatus.objects.filter(id=status.id)
+                .exclude(last_active__gt=freq_offset)
+                .update(last_active=now)
+            )
+
+            if not updated:
+                return
+
+            notification_uuid = str(uuid.uuid4())
+            groupevent = group_to_groupevent[group]
+            rule_fire_history = history.record(rule, group, groupevent.event_id, notification_uuid)
+            safe_execute(
+                activate_downstream_actions, rule, groupevent, notification_uuid, rule_fire_history
+            )
```

Comment:
I think the `safe_execute` is supposed to be on the return value of `activate_downstream_actions`. It 
```suggestion
            for callback, futures in activate_downstream_actions(rule, groupevent, notification_uuid, rule_fire_history).values():
                        safe_execute(callback, groupevent, futures, _with_transaction=False)
```